pkg/api/validate/dynamic/dynamic.go

```diff
@@ -590,34 +598,76 @@ func (dv *dynamic) ValidateSubnets(ctx context.Context, oc *api.OpenShiftCluster
 			)
 		}
 
-		if oc.Properties.ProvisioningState == api.ProvisioningStateCreating {
-			if ss.SubnetPropertiesFormat != nil &&
-				ss.SubnetPropertiesFormat.NetworkSecurityGroup != nil {
-				expectedNsgID, err := subnet.NetworkSecurityGroupID(oc, s.ID)
+		subnetByID[s.ID] = ss
+	}
+	return subnetByID, nil
+}
+
+// checkByoNSG checks whether all the subnets have or don't have NSG attached.
+// when the BYONsg feature flag is on and only some of the subnets are attached with an NSG,
+// it returns an error.  If none of the subnets is attached, it's no longer BYO NSG and the
+// cluster installation process should fall back to using the managed nsg.
+func (dv *dynamic) checkByoNSG(ctx context.Context, subnetByID map[string]*mgmtnetwork.Subnet) (bool, error) {
+	if !dv.byoNSG {
+		return dv.byoNSG, nil // not applicable if the flag is not set
+	}
+
+	noNSGAttached := 0
+	allSubnetsAreAttached := len(subnetByID)
+
+	var attached int
+	for _, subnet := range subnetByID {
+		if subnet.NetworkSecurityGroup != nil && subnet.NetworkSecurityGroup.ID != nil {
+			attached++
+		}
+	}
+	if attached == allSubnetsAreAttached {
+		dv.log.Info("all subnets are attached, BYO NSG")
+		return true, nil // correct setup by customer
+	}
+	if attached == noNSGAttached {
+		dv.log.Info("no subnets are attached, no longer BYO NSG")
+		return false, nil
+	}
+
+	dv.log.Info("BYO NSG: not all subnets are attached")
+	return dv.byoNSG, fmt.Errorf(errMsgNSGNotProperlyAttached)
+}
+
+func (dv *dynamic) ValidateSubnets(ctx context.Context, oc *api.OpenShiftCluster, subnets []Subnet) error {
+	subnetByID, err := dv.createSubnetMapByID(ctx, subnets)
+	if err != nil {
+		return err
+	}
+
+	dv.byoNSG, err = dv.checkByoNSG(ctx, subnetByID)
+	if err != nil {
+		return err
+	}
+
+	for _, s := range subnets {
+		ss := subnetByID[s.ID]
+
+		switch oc.Properties.ProvisioningState {
+		case api.ProvisioningStateCreating:
+			if nsgExists(ss.SubnetPropertiesFormat) && !dv.byoNSG {
+				expectedNsgID, err := getRecordedNsgID(oc, s.ID)
 				if err != nil {
 					return err
 				}
-
-				if !strings.EqualFold(*ss.SubnetPropertiesFormat.NetworkSecurityGroup.ID, expectedNsgID) {
-					return api.NewCloudError(
-						http.StatusBadRequest,
-						api.CloudErrorCodeInvalidLinkedVNet,
-						s.Path,
-						"The provided subnet '%s' is invalid: must not have a network security group attached.",
-						s.ID,
-					)
+				if !isTheSameNSG(*ss.SubnetPropertiesFormat.NetworkSecurityGroup.ID, expectedNsgID) {
```

Comment:
Nope, this is not the same logic as those lines.  It's actually the current logic that rejects the request when it sees if an NSG is attached to either of the subnets.  I changed these lines... kinda for my understanding of how the previous code works (not sure if I'd call it a refactor because it's not like it makes things better).

The only real change that I made is on the line 652 to skip this check when the subscription feature flag is on.